pkg/infra/provider.go

```diff
@@ -0,0 +1,288 @@
+	awsbootstrapv1 "sigs.k8s.io/cluster-api-provider-aws/v2/cmd/clusterawsadm/api/bootstrap/v1beta1"
+	capabootstrap "sigs.k8s.io/cluster-api-provider-aws/v2/cmd/clusterawsadm/cloudformation/bootstrap"
+	cloudformation "sigs.k8s.io/cluster-api-provider-aws/v2/cmd/clusterawsadm/cloudformation/service"
+	"sigs.k8s.io/cluster-api-provider-aws/v2/util/system"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	infrav1 "kurator.dev/kurator/pkg/apis/cluster/v1alpha1"
+	"kurator.dev/kurator/pkg/infra/scope"
+	"kurator.dev/kurator/pkg/infra/template"
+	"kurator.dev/kurator/pkg/infra/util"
+	"kurator.dev/kurator/pkg/typemeta"
+)
+
+type Provider interface {
+	Reconcile(ctx context.Context, cluster *infrav1.Cluster) error
+	Clean(ctx context.Context, cluster *infrav1.Cluster) error
+}
+
+func NewProvider(kube client.Client, scope *scope.Cluster) Provider {
+	return &AWSProvider{
+		Kube:  kube,
+		scope: scope,
+	}
+}
+
+var _ Provider = &AWSProvider{}
+
+type AWSProvider struct {
+	Kube  client.Client
+	scope *scope.Cluster
+}
+
+func (p *AWSProvider) Reconcile(ctx context.Context, infraCluster *infrav1.Cluster) error {
+	clusterCreds, err := p.reconcileAWSCreds(ctx)
+	if err != nil {
+		return errors.Wrapf(err, "failed to reconcile cluster credentials")
+	}
+
+	if err := p.reconcileAWSIAMProfile(ctx, clusterCreds); err != nil {
+		return errors.Wrapf(err, "failed to reconcile IAM profile")
+	}
+
+	// TODO: ensure IRSA
+
+	if _, err := p.reconcileAWSClusterAPIResources(ctx, clusterCreds); err != nil {
+		return errors.Wrapf(err, "failed to reconcile Cluster API resources")
+	}
+
+	// TODO: update VPC.Name if needed
+
+	return nil
+}
+
+func (p *AWSProvider) Clean(ctx context.Context, cluster *infrav1.Cluster) error {
+	// AWSClusterStaticIdentitySpec is not namespaced, so we need to delete the identity by listing all of them matching the cluster labels
+	awsIdentities := &awsinfrav1.AWSClusterStaticIdentityList{}
+	if err := p.Kube.List(ctx, awsIdentities, util.ClusterMatchingLabels(cluster)); err != nil {
+		return errors.Wrapf(err, "failed to list AWSClusterStaticIdentity")
+	}
+
+	for _, identity := range awsIdentities.Items {
+		if err := p.Kube.Delete(ctx, &identity); err != nil {
+			if apierrors.IsNotFound(err) {
+				continue
+			}
+
+			return errors.Wrapf(err, "failed to delete AWSClusterStaticIdentity %s", identity.Name)
+		}
+	}
+
+	return nil
+}
+
+func (p *AWSProvider) reconcileAWSClusterAPIResources(ctx context.Context, clusterCreds *corev1.Secret) (ctrl.Result, error) {
+	_, err := p.reconcileAWSCluster(ctx)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to reconcile AWSCluster %s/%s", p.scope.Namespace, p.scope.Name)
+	}
+
+	b, err := template.RenderClusterAPIForAWS(p.scope)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to render Cluster API resources")
+	}
+	if _, err := util.PatchResources(b); err != nil {
```

Comment:
Is this idempotent?

Reply:
it should, that's why we have speical logic for `AWSCluster`.